File: Sources/UsageWidget/Providers/Codex/CodexProvider.swift
import Foundation
import AppKit

/// Codex provider: loopback OAuth sign-in (127.0.0.1:1455), token refresh, and usage fetch.
actor CodexProvider: UsageProvider {
    nonisolated let id = ProviderID.codex
    nonisolated let displayName = "Codex"
    nonisolated let iconSystemName = "chevron.left.forwardslash.chevron.right"
    nonisolated let accentHex = "#10A37F"
    nonisolated let defaultPollInterval: TimeInterval = 120
    nonisolated let minimumPollInterval: TimeInterval = 60

    private let tokens: TokenStore
    private let oauth = CodexOAuthClient()
    private let fetcher = CodexUsageFetcher()

    init(tokens: TokenStore) {
        self.tokens = tokens
    }

    func authState() async -> ProviderAuthState {
        await tokens.token(for: id) != nil ? .signedIn : .signedOut
    }

    func startSignIn() async throws -> SignInContinuation {
        let pkce = PKCE.generate()
        let server = LoopbackOAuthServer(port: CodexOAuthClient.loopbackPort)
        let url = oauth.makeAuthorizeURL(pkce: pkce)
        await openInBrowser(url)

        let result: LoopbackResult
        do {
            result = try await server.waitForCallback()
        } catch {
            server.stop()
            throw error
        }
        guard result.state == pkce.state else { throw OAuthError.stateMismatch }

        let token = try await oauth.exchange(code: result.code, pkce: pkce)
        await tokens.save(token, for: id)
        return .completed
    }

    func signOut() async {
        await tokens.clear(id)
    }

    func fetchUsage() async throws -> ProviderUsage {
        guard var token = await tokens.token(for: id) else { throw ProviderError.notSignedIn }
        if token.needsRefresh() {
            token = try await refreshAndStore(token)
        }
        do {
            return try await fetcher.fetch(accessToken: token.accessToken, accountId: token.accountId, planFallback: token.planType)
        } catch ProviderError.unauthorized {
            let refreshed = try await refreshAndStore(token)
            return try await fetcher.fetch(accessToken: refreshed.accessToken, accountId: refreshed.accountId, planFallback: refreshed.planType)
        }
    }

    private func refreshAndStore(_ token: OAuthToken) async throws -> OAuthToken {
        do {
            let refreshed = try await oauth.refresh(token)
            await tokens.save(refreshed, for: id)
            return refreshed
        } catch {
            await tokens.clear(id)
            throw ProviderError.unauthorized
        }
    }

    @MainActor
    private func openInBrowser(_ url: URL) {
        NSWorkspace.shared.open(url)
    }
}

File: windows/src/Core/Providers/CodexProvider.cs
using LLMUsageWidget.Core.Auth;
using LLMUsageWidget.Core.Domain;

namespace LLMUsageWidget.Core.Providers;

/// <summary>Codex provider: loopback OAuth (completes inside sign-in), token refresh, and usage fetch.</summary>
public sealed class CodexProvider : IUsageProvider
{
    public ProviderId Id => ProviderId.Codex;
    public string DisplayName => "Codex";
    public string AccentHex => "#10A37F";
    public TimeSpan DefaultPollInterval => TimeSpan.FromSeconds(120);
    public TimeSpan MinimumPollInterval => TimeSpan.FromSeconds(60);

    private readonly ITokenStore _tokens;
    private readonly Action<Uri> _openBrowser;
    private readonly CodexOAuthClient _oauth = new();
    private readonly CodexUsageFetcher _fetcher = new();

    public CodexProvider(ITokenStore tokens, Action<Uri> openBrowser)
    {
        _tokens = tokens;
        _openBrowser = openBrowser;
    }

    public Task<ProviderAuthState> AuthStateAsync() =>
        Task.FromResult(_tokens.Get(Id) is not null ? ProviderAuthState.SignedIn : ProviderAuthState.SignedOut);

    public async Task<SignInFlow> StartSignInAsync()
    {
        var pkce = Pkce.Generate();
        _openBrowser(_oauth.MakeAuthorizeUrl(pkce));
        string code = await LoopbackServer.WaitForCodeAsync(CodexOAuthClient.LoopbackPort, pkce.State, TimeSpan.FromMinutes(5));
        var token = await _oauth.ExchangeAsync(code, pkce);
        _tokens.Save(Id, token);
        return new SignInFlow.Completed();
    }

    public Task SignOutAsync()
    {
        _tokens.Clear(Id);
        return Task.CompletedTask;
    }

    public async Task<ProviderUsage> FetchUsageAsync()
    {
        var token = _tokens.Get(Id) ?? throw ProviderException.NotSignedIn();
        if (token.NeedsRefresh()) token = await RefreshAndStoreAsync(token);
        try
        {
            return await _fetcher.FetchAsync(token.AccessToken, token.AccountId, token.PlanType);
        }
        catch (ProviderException ex) when (ex.Kind == ProviderErrorKind.Unauthorized)
        {
            var refreshed = await RefreshAndStoreAsync(token);
            return await _fetcher.FetchAsync(refreshed.AccessToken, refreshed.AccountId, refreshed.PlanType);
        }
    }

    private async Task<OAuthToken> RefreshAndStoreAsync(OAuthToken token)
    {
        try
        {
            var refreshed = await _oauth.RefreshAsync(token);
            _tokens.Save(Id, refreshed);
            return refreshed;
        }
        catch
        {
            _tokens.Clear(Id);
            throw ProviderException.Unauthorized();
        }
    }
}
